### PR TITLE
Stabilize `type_alias_enum_variants` in Rust 1.37.0

### LINKED ISSUE #52118: Using variant on `self` type with path `Self::Variant` causes subpar diagnostic
The snippet below fails to compile with the following messages of each `enum` variant:

```rust
pub enum State {
    Init,
    NewProject,
    NewTask,
    NewEvent,
}

impl State {
    pub fn commands<'a>(&self) -> usize {
        match self {
            Self::Init => 1,
            Self::NewProject | Self::NewTask | Self::NewEvent => 2,
            _ => 0,
        }
    }
}

fn main() {}
```

```
error[E0599]: no variant named `Init` found for type `State` in the current scope
  --> src/main.rs:11:13
   |
1  | pub enum State {
   | -------------- variant `Init` not found here
...
11 |             Self::Init => 1,
   |             ^^^^^^^^^^ variant not found in `State`
   |
   = note: did you mean `variant::Init`?
```

The solution is to replace `Self::Init` with explicit type `State::Init`. But the issues are:

- Error E0599 doesn't explain the actual problem (Use of `Self` instead of the explicit type)
- The message `no variant named "Init" found for type "State" in the current scope` is misleading
- `note: did you mean "variant::Init"?` is also a little misleading. A phrase like `note: replace "Self::Init" with "State::Init"` is much clearer

And finally:
- Why isn't `Self::` allowed inside `impl MyEnum` blocks?

### PR DESCRIPTION
**Pull request by Centril**
Stabilize `#![feature(type_alias_enum_variants)]` which allows type-relative resolution with highest priority to `enum` variants in both expression and pattern contexts. For example, you may now write:

```rust
enum Option<T> {
    None,
    Some(T),
}

type OptAlias<T> = Option<T>;

fn work_on_alias(x: Option<u8>) -> u8 {
    match x {
        OptAlias::Some(y) => y + 1,
        OptAlias::None => 0,
    }
}
```

Closes https://github.com/rust-lang/rfcs/issues/2218
Closes https://github.com/rust-lang/rust/issues/52118

r? @petrochenkov

### CODE REVIEW

**Comment by Centril** [comment]
# Stabilization proposal

I propose that we stabilize `#![feature(type_alias_enum_variants)]`.

@rfcbot merge

Tracking issue: https://github.com/rust-lang/rust/issues/49683
Version target: 1.37 (2019-07-04 => beta, 2019-08-15 => stable).

## What is stabilized

### User guide

#### Basics

You can now refer to an `enum` variant through a type alias in expression and pattern contexts. For example, the following becomes legal:

```rust
enum Option<T> {
    None,
    Some(T),
}

type OptAlias<T> = Option<T>;

fn work_on_alias(x: Option<u8>) -> u8 {
    match x {
        OptAlias::Some(y) => y + 1,
        OptAlias::None => 0,
    }
}
```

You can also *fix* the type arguments in a type alias like so:

```rust
type FixedAlias = Option<u8>;

fn work_on_fixed_alias(x: Option<u8>) -> u8 {
    match x {
        FixedAlias::Some(y) => y + 1,
        FixedAlias::None => 0,
    }
}
```

#### Generics

To apply a type argument to `OptAlias` and refer to one of the variants, you may write:

```rust
OptAlias::<MyType>::None // OK.
Option::<MyType>::None // Also works.
```

However, unlike when referring directly to an `enum` variant, you cannot write:

```rust
OptAlias::None::<MyType> // Not OK!
Option::None::<MyType> // OK for backwards compatibility.
```

#### `Self` as a type alias

You will recall that `Self` in an implementation acts like a type alias referring to the implementing type. With this in mind, and the newfound ability to resolve to `enum` variants through type alias, we can now also refer to `enum` variants through `Self`. For example:

```rust
impl<A> Option<A> {
    fn map<B>(self, transform: impl FnOnce(A) -> B) -> Option<B> {
        match self {
            Self::Some(a) => {
                Option::Some(transform(a))
            }
            Self::None => {
                Option::None
            }
        }
    }
}
```

Notice that we are using `Option::None` to produce a value of `Option<B>`.
Instead, you might have expected (1):

```rust
        match self {
            Self::Some(a) => {
                Self::Some(transform(a))
            }
            Self::None => {
                Self::None
            }
        }
```

However, this is not allowed because `Self` refers to the *type* `Option<A>` which is not the same as `Option<B>`. The compiler will therefore reject (1) with:

```rust
[...]

error[E0308]: mismatched types
  --> src/lib.rs:14:17
   |
8  |     fn map<B>(self, arrow: impl FnOnce(A) -> B) -> Option<B> {
   |                                                    --------- expected `Option<B>` because of return type
...
14 |                 Self::None
   |                 ^^^^^^^^^^ expected type parameter, found a different type parameter
   |
   = note: expected type `Option<B>`
              found type `Option<A>`
```

#### Type relative resolution

When you write `Foo::Bar` to refer to an associated item, this is actually sugar for `<Foo<..>>::Bar`. What we stabilize here is the ability to resolve to `enum` variants through type-relative resolution. Therefore, you can write `<MyEnum<_>>::Variant`, `<MyAlias>::Variant`, or even refer to `enum` variants through an associated type projection as in the following example:

```rust
enum Enum {
    Variant,
}

trait Proto {
    type Assoc;
}

struct Unit;

impl Proto for Unit {
    type Assoc = Enum;
}

fn foo() {
    let x = < <Unit as Proto>::Assoc >::Variant;
}
```

### Motivation

Stabilizing `#![feature(type_alias_enum_variants)]` has a few notable justifications:

1. "It is a bug-fix"

   [The general belief](https://github.com/rust-lang/rust/issues/26264#issuecomment-250603224) of the language team has been that [variants are inherent associated items](https://github.com/rust-lang/rust/issues/26264#issuecomment-250603946). As it is already possible to refer to inherent associated items through type aliases, this capability ought to extend consistently to all such items and that includes `enum` variants.

2. Fewer surprises and better ergonomics

   The inability to refer to `enum` variants through type aliases is a long standing issue dating back at least to [2015](https://github.com/rust-lang/rust/issues/26264). This is an issue because [many people](https://github.com/rust-lang/rust/issues/26264#issuecomment-130493681) expect it to work and are surprised when it does not. Reducing that surprise would be helpful towards a language that is easier to learn.

   It is of [particular](https://github.com/rust-lang/rust/issues/49636) note that `Self::Variant` [does not work](https://github.com/rust-lang/rust/issues/55157) while many users expect it to.

   Over the years, we've made `Self` work in more and more places. For example, you may use `Self` to construct and pattern match on unit, tuple, and braced structs inside an implementation. By allowing `Self::Variant` to work, we can complete more of this story. This is also an ergonomics benefit as you can just use:

   ```rust
   match self {
       Self::Foo(..) => foo_logic(),
       Self::Bar(..) => bar_logic(),
       ...
   }
   ```

   without having to care about what the specific name of `Self` is when writing or reading code.

4. Abstraction

   One key advantage to type aliases is that you may apply certain types to a generic type and then then introduce a conceptual name for your generic type + type arguments. However, as `enum` variants currently cannot be used through type aliases, you lose out on parts of this advantage for `enum`s. To stabilize `type_alias_enum_variants` would rectify this.

### Reference guide

In the value namespace, type-relative resolution to an `enum` variant is permitted.
This applies to expressions and patterns alike.

1. Type relative resolution

   - Let `E` denote an `enum` definition with variants `V0` to `Vn`.

   - Let `F` denote a type expression normalizing to `E` with generic substitutions `Si`.
   
   - Then a path `<F>::Vj` may resolve, with highest priority, in the value namespace, to the variant `Vj`.

2. Desugaring to type-relative resolution *(pre-existing on stable)*

   - Let `X` be an associated item of `G`.

   - A path `G::X` is equivalent to `<G<..>>::X` where `..` denotes as many inference variables as `G` needs if it is generic.

   - A path `G::<A0, .., An>::X` is equivalent to `<G::<A0, .., An>>::X`.

## Divergences from RFC

There are two principle divergences from the RFC:

### `foo::Some::<u8>` is not allowed

The RFC states that this should be allowed:

```rust
type Alias<T> = Option<T>;

mod foo {
    pub use Alias::Some;
}

fn bar() {
    foo::Some::<u8>;
}
```

However, the current implementation does not allow the example above.

The feature description of "type alias enum variants" is [somewhat of a misnomer](https://github.com/rust-lang/rust/issues/49683#issuecomment-466631781). What actually is implemented is *"type-relative enum variants"* or *"variants as associated items"*. As @petrochenkov notes:

> When writing `<TYPE>::NAME` you are asking to search for an associated item named `NAME` on type `TYPE`, for variants this kind of requests is feature gated.

Further, @eddyb [notes](https://github.com/rust-lang/rust/issues/49683#issuecomment-458821419) that:

> For `use E2::*;` to work, it would have to be sugar for `use <E2>::*;`, and involve the typesystem in name resolution (if any generic arguments are involved, you would have to transport them faithfully), which isn't really doable atm.
> This is a more general problem, in that `use Foo::...;` won't import associated consts/types/methods from types/traits.

### `Alias::Variant::<TypeArgs>` is not allowed but `Alias::<TypeArgs>::Variant` is

The RFC specified that:

```rust
Enum::Variant::<TypeArgs> // OK.
Enum::<TypeArgs>::Variant // ERROR.
Alias::Variant::<TypeArgs> // OK.
Alias::<TypeArgs>::Variant // ERROR.
```

Instead, we have implemented:

```rust
Enum::Variant::<TypeArgs> // OK (1).
Enum::<TypeArgs>::Variant // OK.
Alias::Variant::<TypeArgs> // ERROR.
Alias::<TypeArgs>::Variant // OK.
```

There are a few reasons why we should diverge in this manner:

1. `Enum::<TypeArgs>::Variant` is already allowed on stable.

2. [We understand](https://github.com/rust-lang/rust/issues/26264#issuecomment-250603946) `enum` variants to be inherent associated items and they are resolved as such. Therefore, as @petrochenkov [notes](https://github.com/rust-lang/rfcs/pull/2338#issuecomment-367456633):

   > `Ty<Args1>::AssocItem<Args2>` or `Trait<Args1>::AssocItem<Args2>` means that `Args1` are passed to the parent item and `Args2` are passed to the associated item. No argument transfer between segments happen, arguments are always passed to their proper segment.
   >
   > For variants that would mean `EnumAlias<Args1>::Variant` and `Enum<Args1>::Variant` being the correct way pass generic arguments (variants don't have their own generic parameters (yet) so nothing is passed to them).

   Indeed, while variants do not have their own generic parameters yet, they could very well have if some form of [*generalized algebraic data types* (GADTs)](https://en.wikipedia.org/wiki/Generalized_algebraic_data_type) were to be implemented..

   @petrochenkov further notes that:

   > `Enum::Variant<Args1>` being legal and meaning `Enum<Args1>::Variant` exists for historical reasons and is, IMO, a wart. We should not extend it on type aliases (but we should keep it for backward compatibility of course). More details in https://internals.rust-lang.org/t/pre-rfc-enum-variants-through-type-aliases/6433/5.
   > So, I think we should go the `EnumAlias<Args1>::Variant` way instead of the scheme proposed in the RFC.

   I (@Centril) agree with this assessment. As future work we may consider deprecating `EnumAlias<Args1>::Variant` from the language and eventually remove it in an edition.

3. @eddyb [believes](https://github.com/rust-lang/rust/pull/56225#issuecomment-443616173) that `Alias::Variant::<TypeArgs>` is generally unimplementable.

## Resolved questions

Enum variants are [implemented](https://github.com/rust-lang/rust/pull/57501) to resolve with highest priority when there is ambigutiy. There are numerous reasons why this is the right choice as noted by @petrochenkov:

> - If variants (and their constructors) are treated as associated items, then they are obviously inherent associated items since they don't come from traits.
> - Inherent associated items have higher priority during resolution than associated items from traits.
> - The reason is that there is a way to disambiguate in favor of trait items (<Type as Trait>::Ambiguous), but there's no way to disambiguate in favor of inherent items, so they become unusable in case of ambiguities if they have low priority.
> - It's technically problematic to fallback from associated types to anything until lazy normalization (?) is implemented.

## Potential concerns

Should https://github.com/rust-lang/rfcs/pull/2593 be accepted, `enum` variants would exist in the type namespace as well. If that should happen, and given that `enum` variants resolve with highest priority for the reasons aforementioned, the following code could break:

```rust
enum E { V }

trait T {
    type V;
    const C: Self::V;
}

impl T for E {
    type V = u8;
    const C: Self::V = 0;
          // ------- currently refers to `type V`
          //         with RFC 2593 it refers to the *type* `E::V`.
}
```

The potential for breakage only exists should both the current proposal be stabilized as well as RFC 2593 at a later time. To prevent instantaneous breakage, a future compatibility lint [`ambiguous_associated_items`](https://github.com/rust-lang/rust/issues/57644) is currently deny-by-default on the beta compiler. Given the code above, a user will see:

```rust
error: ambiguous associated item
  --> src/lib.rs:10:14
   |
10 |     const C: Self::V = 0;
   |              ^^^^^^^ help: use fully-qualified syntax: `<E as Trait>::V`
   |
   = note: #[deny(ambiguous_associated_items)] on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #57644 <https://github.com/rust-lang/rust/issues/57644>
note: `V` could refer to variant defined here
  --> src/lib.rs:1:10
   |
1  | enum E { V }
   |          ^
note: `V` could also refer to associated type defined here
  --> src/lib.rs:4:5
   |
4  |     type V;
   |     ^^^^^^^
```

The lint should give us peace of mind that RFC 2593 can be implemented should we so desire.

As for the existing regressions, @varkor [notes](https://github.com/rust-lang/rust/pull/59928#issuecomment-488664195) that there are only 4 remaining regressions of which all have open PR issues to fix them. Moreover, none of the 4 crates have reverse dependencies and all appear to be abandonware.

## Tests

The tests can be primarily seen in the PR itself. Here are some of them:

- [Feature gate test being removed](https://github.com/rust-lang/rust/blob/55cee44671ecb0869cc7fbac0ad62f6236893d06/src/test/ui/feature-gates/feature-gate-type_alias_enum_variants.rs)

- [Unstable book entry being removed](https://github.com/rust-lang/rust/blob/55cee44671ecb0869cc7fbac0ad62f6236893d06/src/doc/unstable-book/src/language-features/type-alias-enum-variants.md)

- [It is possible to resolve, in the value namespace, to an `enum` variant through a type alias, including `Self`.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/type-alias-enum-variants-pass.rs) The test accounts both for expression and pattern contexts. The test also checks that `<TypeAlias>::Variant` also works.

- [`Alias::None::<u8>` is rejected.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/no-type-application-on-aliased-enum-variant.rs)

- [Creating/matching on an enum variant through an alias with the wrong braced/unit form is caught as an error.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/incorrect-variant-form-through-alias-caught.rs)

- [The `ambiguous_associated_items` lint is deny-by-default and correctly catches the ambiguous situation.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/enum-variant-priority-lint-ambiguous_associated_items.rs) The test elaborates on the problematic scenario in relation to https://github.com/rust-lang/rfcs/pull/2593.

- [An `enum` variant is resolved, in the value namespace, with higher priority than other inherent items when there is a conflict.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/enum-variant-priority-higher-than-other-inherent.rs)

- [Applied type arguments of enums, and aliases to them, are respected.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/enum-variant-generic-args.rs) We also check that the variant to a type-aliased enum cannot be type applied whether that alias is generic or monomorphic.

- [The implementation is well behaved wrt. generics.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/enum-variant-generic-args-pass.rs)

- [The compiler will resolve `<E>::V` to the variant `V` in the type namespace but will reject this because `enum` variants do not exist in the type namespace.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/resolve-to-enum-variant-in-type-namespace-and-error.rs)

- [Path patterns correctly infer a type when referring to a generic `enum` via an alias.](https://github.com/rust-lang/rust/blob/46f405ec4d7c6bf16fc2eaafe7541019f1da2996/src/test/ui/type-alias-enum-variants/type-alias-enum-variants-pass.rs)

## History

- On 2015-06-13, @AngelOfSol raises https://github.com/rust-lang/rust/issues/26264 re. `EnumAlias::Variant` not working.

  - @chris-morgan [notes in a comment](https://github.com/rust-lang/rust/issues/26264#issuecomment-130493681), that it is reasonable to expect this to work.

  - The discussion goes on and many people concur.

  - The issue gets cited many times over the years. Many other duplicate issues such as https://github.com/rust-lang/rust/issues/28556, https://github.com/rust-lang/rust/issues/31355, and https://github.com/rust-lang/rust/issues/31168 are made.

  - On 2016-09-29, @nikomatsakis [notes in a comment](https://github.com/rust-lang/rust/issues/26264#issuecomment-250603224) that consider the current behaviour to be a bug. Moreover, @nikomatsakis thinks of variants as associated items.

  - On 2016-09-30, @aturon, on behalf of the language team of the date, [concurs](https://github.com/rust-lang/rust/issues/26264#issuecomment-250603946) with @nikomatsakis's assessment.

- On 2016-01-25, @jseyfried [makes a PR](https://github.com/rust-lang/rust/pull/31179) to support the concept.

  - Some thornier questions are raised by @petrochenkov in https://github.com/rust-lang/rust/pull/31179#issuecomment-174449078.

  - The language team discusses the PR. On their behalf, @aturon [notes](https://github.com/rust-lang/rust/pull/31179#issuecomment-177035852) on 2019-01-30 that that an RFC is necessary and the PR is closed.

- On 2019-12-28, @djc makes [a Pre-RFC](https://internals.rust-lang.org/t/pre-rfc-enum-variants-through-type-aliases/6433) raising the issue again. After some discussion...

- On 2018-04-05, @djc proposes [RFC 2338](https://github.com/rust-lang/rfcs/pull/2338), *"Type alias enum variants"*.

  - There is [some](https://github.com/rust-lang/rfcs/pull/2338#issuecomment-367456633) discussion [about](https://github.com/rust-lang/rfcs/pull/2338#issuecomment-367775440) type argument transfer, i.e. `Alias::Variant::<Args..>` transferring to `Alias::<Args..>::Variant`. The issue does not appear discussed in depth by the language team. In the tracking issue, @petrochenkov later [reiterates](https://github.com/rust-lang/rust/issues/49683#issuecomment-441482546) that we should consider the argument transfer and keep `None::<T>` for compatibility but not to expand the hack further.

  - Later, @alexreg [notes](https://github.com/rust-lang/rfcs/pull/2338#discussion_r236116988) that as `mod foo { pub use Alias::Some; }` should not work initially and @petrochenkov agrees [noting](https://github.com/rust-lang/rfcs/pull/2338#discussion_r236200205) that `Alias::Some` is type-relative in terms of resolution and needs type-checking to determine what `Alias` refers to which isn't currently possible due to phase separation.

- On 2018-04-04, [RFC 2338 is merged](https://github.com/rust-lang/rfcs/pull/2338#issuecomment-378863845) and the [tracking issue](https://github.com/rust-lang/rust/issues/49683) is made.

- On 2018-12-30, the RFC is partially implemented by @alexreg in https://github.com/rust-lang/rust/pull/56225. @petrochenkov reviews the PR with @eddyb chipping in.

  - @djc [raises concerns](https://github.com/rust-lang/rust/pull/56225#issuecomment-443614942) re. divergences in the PR relative to the RFC.

  - In response, @eddyb [explains](https://github.com/rust-lang/rust/pull/56225#issuecomment-443616173) that the RFC as stated is unimplementable and attributes this to an oversight during the RFC. @eddyb also concurs with @petrochenkov's notes during the RFC that `Alias::Variant` is type-relative in the sense that it is sugar for `<Alias>::Variant`.

  - @alexreg [notes](https://github.com/rust-lang/rust/pull/56225#issuecomment-443782997) that they have enabled `Enum::<...>::Variant`; this is becomes stable in 1.33.0.

- On 2019-01-19, https://github.com/rust-lang/rust/pull/57501 by @petrochenkov and reviewed by @alexreg is merged. The PR changes variants to resolve with lowest priority, as implemented in https://github.com/rust-lang/rust/pull/56225, to being resolved with the highest priority. Motivation for this change is included in the PR description.

  - A crater run is done. [Some possible future regressions are found.](https://github.com/rust-lang/rust/pull/57501#issuecomment-453730114)
    To preserve future compatibility with the combination of `#![feature(type_alias_enum_variants)]` and https://github.com/rust-lang/rfcs/pull/2593, a C-future compatibility lint `ambiguous_associated_items` is made.

- On 2019-04-12, at the request of @Centril, @petrochenkov writes PR https://github.com/rust-lang/rust/pull/59928 to change `ambiguous_associated_items` to be deny-by-default. The PR, merged on 2019-05-03, is reviewed by @varkor.

   - @Centril files PRs and issues against the affected crates to fix the regressions.

   - As @varkor [notes](https://github.com/rust-lang/rust/pull/59928#issuecomment-488664195), before the PR is merged, there are only 4 remaining regressions of which all have open PR issues to fix them. Moreover, none of the 4 crates have reverse dependencies and all appear to be abandonware.

- On 2019-06-15, https://github.com/rust-lang/rust/pull/61825, written by @Centril and reviewed by @petrochenkov, was merged. The PR fixed an issue where a path pattern, e.g. `OptAlias::None`, referring to a generic `enum`'s variant through a type alias failed to infer the type argument applied to said `enum` even though the expectation was set by the `match` scrutinee expression.

- This PR stabilizes `type_alias_enum_variants`, hardens the tests, and documents all of them.

## Possible future work

Some possible work in the future includes:

1. Deprecating `Enum::Variant::<Args..>` and `Variant::<Args..>`.

2. Supporting importing items in traits and inherent implementations. For example:

    - Traits:

      ```rust
      use Default::default;
      ```

    - Inherent items:

      ```rust
      struct A;

      impl A {
          fn foo() {}
      }

      use A::foo;
      ```

File: src/test/ui/type-alias-enum-variants/enum-variant-generic-args-pass.rs [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
+// Check that resolving, in the value namespace, to an `enum` variant
+// through a type alias is well behaved in the presence of generics.
+// We check for situations with:
+// 1. a generic type `Alias<T>`, we can type-apply `Alias` when referring to a variant.
+// 2. a monotype `AliasFixed` of generic `Enum<T>`, we can refer to variants
+//    and the type-application of `T` in `AliasFixed` is kept.
 
 #![allow(irrefutable_let_patterns)]
 
 #[allow(dead_code)]
```
Comment (petrochenkov) [inline src/test/ui/type-alias-enum-variants/enum-variant-generic-args-pass.rs]: Lints from the `unused` group are `allow`ed by default in UI tests.

File: src/test/ui/type-alias-enum-variants/incorrect-variant-form-through-alias-caught.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// ignore-tidy-linelength
+
+// Check that creating/matching on an enum variant through an alias with
+// the wrong braced/unit form is caught as an error.
+
+#![allow(unreachable_code)]
```
Comment (petrochenkov) [inline src/test/ui/type-alias-enum-variants/incorrect-variant-form-through-alias-caught.rs]: Lints from the `unused` group are `allow`ed by default in UI tests.

**Comment by petrochenkov** [comment]
>In the value namespace, type-relative resolution to an enum variant is permitted.

The type-relative variant resolution works in type namespace as well.
The case in which it happens without errors is struct expressions/patterns:
```
<Enum>::Variant { ... } // `Variant` is searched in type namespace
```

>there is a way to disambiguate in favor of trait items (::Ambiguous)

Markdown has broken during copy-pasting.

**Comment by petrochenkov** [comment]
Implementation and report LGMT, marking as waiting-on-team.

File: src/test/ui/type-alias-enum-variants/enum-variant-generic-args-pass.rs (hunk 1 above)
Comment (Centril) [reply src/test/ui/type-alias-enum-variants/enum-variant-generic-args-pass.rs]: Fixed in https://github.com/rust-lang/rust/pull/61682/commits/57e6869a6be56cd75e382deb313fb5f4d1bb1fda.

File: src/test/ui/type-alias-enum-variants/incorrect-variant-form-through-alias-caught.rs (hunk 2 above)
Comment (Centril) [reply src/test/ui/type-alias-enum-variants/incorrect-variant-form-through-alias-caught.rs]: Fixed in https://github.com/rust-lang/rust/pull/61682/commits/57e6869a6be56cd75e382deb313fb5f4d1bb1fda.

**Comment by Centril** [comment]
N.B. I've checked @aturon's box since they are on leave.

**Comment by glaebhoerl** [comment]
> I (@Centril) agree with this assessment. As future work we may consider deprecating `EnumAlias<Args1>::Variant` from the language and eventually remove it in an edition.

I suspect you meant to write `Enum::Variant<Args1>` here?

**Comment by Centril** [comment]
@rfcbot reviewed

Apparently the bot is slow :P

**Comment by auscompgeek** [comment]
> * On 2019-12-28, @djc makes [a Pre-RFC](https://internals.rust-lang.org/t/pre-rfc-enum-variants-through-type-aliases/6433) raising the issue again.

Have we time travelled?

**Comment by pietroalbini** [comment]
@bors retry

Yielding priority to the Azure switch.

**Pull request merged**